pipeline/id_minter/src/main/scala/uk/ac/wellcome/platform/idminter/database/IdentifiersDao.scala

```diff
@@ -1,82 +1,145 @@
+        LookupResult(
+          foundIdentifiers,
+          notFoundIdentifiers.toList
+        )
+      }
     }
   }
 
-  /* Save an identifier into the database.
-   *
-   * Note that this will copy _all_ the fields on `Identifier`, nulling any
-   * fields which aren't set on `Identifier`.
-   */
-  def saveIdentifier(identifier: Identifier): Try[Any] = {
+  @throws(classOf[InsertError])
+  def saveIdentifiers(ids: List[Identifier]): Try[InsertResult] =
     Try {
+      val values = ids.map(i =>
+        Seq(i.CanonicalId, i.OntologyType, i.SourceSystem, i.SourceId))
       blocking {
-        debug(s"Putting new identifier $identifier")
+        debug(s"Putting new identifier $ids")
         withSQL {
           insert
             .into(identifiers)
             .namedValues(
-              identifiers.column.CanonicalId -> identifier.CanonicalId,
-              identifiers.column.OntologyType -> identifier.OntologyType,
-              identifiers.column.SourceSystem -> identifier.SourceSystem,
-              identifiers.column.SourceId -> identifier.SourceId
+              identifiers.column.CanonicalId -> sqls.?,
+              identifiers.column.OntologyType -> sqls.?,
+              identifiers.column.SourceSystem -> sqls.?,
+              identifiers.column.SourceId -> sqls.?
             )
-        }.update().apply()
+        }.batch(values: _*).apply()
+        InsertResult(ids)
       }
     } recover {
-      case e: SQLIntegrityConstraintViolationException =>
-        warn(
-          s"Unable to insert $identifier because of integrity constraints: ${e.getMessage}")
-        throw IdMinterException(e)
+      case e: BatchUpdateException =>
+        val insertError = getInsertErrorFromException(e, ids)
+        val failedIds = insertError.failed.map(_.SourceId)
+        val succeededIds = insertError.succeeded.map(_.SourceId)
+        error(
+          s"Batch update failed for [$failedIds], succeeded for [$succeededIds]",
+          e)
+        throw insertError
       case e =>
-        error(s"Failed inserting identifier $identifier in database", e)
+        error(s"Failed inserting IDs: [${ids.map(_.SourceId)}]")
         throw e
     }
+
+  private def getInsertErrorFromException(
+    exception: BatchUpdateException,
+    ids: List[Identifier]): InsertError = {
+    val groupedResult = ids.zip(exception.getUpdateCounts).groupBy {
+      case (_, Statement.EXECUTE_FAILED) => Statement.EXECUTE_FAILED
+      case (_, _)                        => Statement.SUCCESS_NO_INFO
+    }
+    val failedIdentifiers =
+      groupedResult.getOrElse(Statement.EXECUTE_FAILED, Nil).map {
+        case (identifier, _) => identifier
+      }
+    val succeededIdentifiers =
+      groupedResult.getOrElse(Statement.SUCCESS_NO_INFO, Nil).map {
+        case (identifier, _) => identifier
+      }
+    InsertError(failedIdentifiers, exception, succeededIdentifiers)
+  }
```

Comment:
FYI, I know very little about SQL errors or the like, so I’m taking it on faith this works correctly. (Comments would be nice!)

Reply:
To be honest I'm doing the same - @alicefuzier did the difficult stuff (ie the SQL) in this PR! If it proves to be a problem we can add some comments in future